File: src/voice_kb/state.py
"""Session state machine.

Pure. Events go in, ``(next_state, command)`` comes out; nothing here touches a
device, the model, or X. The imperative shell in :mod:`voice_kb.app` interprets
the commands.

The state is a closed union, so "recording without a capture start time" or
"transcribing with no audio" cannot be represented.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Final, assert_never


class Phase(Enum):
    """Coarse phase, for the indicator. Derived from the state, never stored."""

    IDLE = auto()
    RECORDING = auto()
    TRANSCRIBING = auto()


# --------------------------------------------------------------------------- states


@dataclass(frozen=True, slots=True)
class Idle:
    """Nothing in flight. The audio stream may still be open for pre-roll."""


@dataclass(frozen=True, slots=True)
class Recording:
    """Audio is accumulating. Either the key is held, or recording is latched."""

    started_at: float
    """``time.monotonic()`` when the key went down."""

    latched: bool = False
    """Whether this recording outlives the key release.

    ``False`` is push-to-talk: the release ends it. ``True`` is a latched
    recording, started with the modifier held, which ignores the release and
    runs until the key is pressed again. The distinction has to be *in the
    state* rather than read off the event that ends it, because the ending
    event is a different one in each mode -- a ``KeyUp`` for push-to-talk, a
    ``KeyDown`` for a latch.
    """


@dataclass(frozen=True, slots=True)
class Transcribing:
    """The key was released; a decode is running off the main thread."""

    started_at: float
    """``time.monotonic()`` when the key went down (not when decoding began)."""

    released_at: float
    """``time.monotonic()`` when the key came up."""

    @property
    def spoken_seconds(self) -> float:
        return self.released_at - self.started_at


type SessionState = Idle | Recording | Transcribing


def phase_of(state: SessionState) -> Phase:
    match state:
        case Idle():
            return Phase.IDLE
        case Recording():
            return Phase.RECORDING
        case Transcribing():
            return Phase.TRANSCRIBING
        case _:
            assert_never(state)


# --------------------------------------------------------------------------- events


@dataclass(frozen=True, slots=True)
class KeyDown:
    at: float

    latch: bool = False
    """The latch modifier was held when the key went down.

    Only meaningful for a press that *starts* a recording. The press that
    ends a latched recording stops it whether or not the modifier was held,
    so the user never has to remember which hand they started with.
    """


@dataclass(frozen=True, slots=True)
class KeyUp:
    at: float


@dataclass(frozen=True, slots=True)
class DecodeFinished:
    """The decode completed, successfully or not."""

    at: float


@dataclass(frozen=True, slots=True)
class Cancelled:
    """User abort (cancel key), or a decode that failed hard."""

    at: float


type Event = KeyDown | KeyUp | DecodeFinished | Cancelled


# ------------------------------------------------------------------------- commands


@dataclass(frozen=True, slots=True)
class Nothing:
    """The event is not meaningful in this state; ignore it.

    This is the common case for key auto-repeat, which X11 delivers as a stream
    of down events while a key is held.
    """


@dataclass(frozen=True, slots=True)
class StartCapture:
    """Begin accumulating audio, seeded with the pre-roll ring buffer."""


@dataclass(frozen=True, slots=True)
class Decode:
    """Stop accumulating and decode what was captured."""

    spoken_seconds: float


@dataclass(frozen=True, slots=True)
class DiscardCapture:
    """Stop accumulating and throw the audio away."""


@dataclass(frozen=True, slots=True)
class AbortDecode:
    """Invalidate the in-flight decode so its result is never injected.

    Without this the shell has no way to distinguish "the decode the user is
    still waiting for" from "the decode the user cancelled two seconds ago",
    and stale text lands in whatever window happens to be focused by the time
    it finishes.
    """


type Command = Nothing | StartCapture | Decode | DiscardCapture | AbortDecode


# ----------------------------------------------------------------------- transition


MIN_HOLD_SECONDS: Final = 0.12
"""Below this, a keypress is a stray tap rather than dictation.

The hotkey is deliberately not grabbed, so the key stays live for everything
else on the system and accidental presses are expected. Without this floor a
30ms brush of the key decodes a fraction of a second of room noise and pastes
whatever the model hallucinates into the focused window.
"""


def step(
    state: SessionState,
    event: Event,
    *,
    min_hold_seconds: float = MIN_HOLD_SECONDS,
) -> tuple[SessionState, Command]:
    """Total transition function.

    Every (state, event) pair is handled. Events that are meaningless in the
    current state return the state unchanged with :class:`Nothing` rather than
    raising -- key auto-repeat and late decode callbacks both rely on this.
    """
    match state, event:
        # -- starting
        case Idle(), KeyDown(at=at, latch=latch):
            return Recording(started_at=at, latched=latch), StartCapture()

        # A press arriving while a previous decode is still running starts a new
        # utterance rather than being dropped. The decode runs on its own thread
        # and is unaffected; its result still gets appended when it lands.
        case Transcribing(), KeyDown(at=at, latch=latch):
            return Recording(started_at=at, latched=latch), StartCapture()

        # -- finishing a latched recording: the *press* ends it, not the release
        #
        # This case has to precede the KeyDown cases above in intent, but it
        # cannot conflict with them: those match Idle and Transcribing, and a
        # latched recording is neither.
        case Recording(started_at=started, latched=True), KeyDown(at=at) if (
            at - started < min_hold_seconds
        ):
            return Idle(), DiscardCapture()

        case Recording(started_at=started, latched=True), KeyDown(at=at):
            return (
                Transcribing(started_at=started, released_at=at),
                Decode(spoken_seconds=at - started),
            )

        # A latched recording ignores the release that started it, and every
        # release after. Without this the very first key-up would end it.
        case Recording(latched=True), KeyUp():
            return state, Nothing()

        # -- finishing push-to-talk
        case Recording(started_at=started), KeyUp(at=at) if at - started < min_hold_seconds:
            return Idle(), DiscardCapture()

        case Recording(started_at=started), KeyUp(at=at):
            return (
                Transcribing(started_at=started, released_at=at),
                Decode(spoken_seconds=at - started),
            )

        # -- aborting
        case Recording(), Cancelled():
            return Idle(), DiscardCapture()

        case Transcribing(), Cancelled():
            return Idle(), AbortDecode()

        # -- settling
        case Transcribing(), DecodeFinished():
            return Idle(), Nothing()

        # -- everything else is noise: auto-repeat, duplicate ups, stale callbacks
        case _:
            return state, Nothing()

File: src/state.rs
//! Pure session transitions. Repeat filtering belongs to the input adapter.
use crate::nvim::IndicatorPhase;
use std::time::{Duration, Instant};

/// Below this a press is a stray tap, not dictation. Deliberately not
/// configurable: it is a property of how a key feels, not of a deployment.
pub const MINIMUM_HOLD: Duration = Duration::from_millis(120);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum State {
    Idle,
    Recording { started: Instant, latched: bool },
    Transcribing { started: Instant, released: Instant },
}

#[derive(Debug, Clone, Copy)]
pub enum Event {
    Down {
        at: Instant,
        latch: bool,
    },
    Up {
        at: Instant,
    },
    /// No watched keyboard can report the hotkey any more — the one holding it
    /// disappeared, or the last one that advertises it did. A recording must
    /// never outlive the key that would end it.
    HotkeyLost {
        at: Instant,
    },
    Finished,
    Cancel,
}

/// Why a capture was thrown away. The user is told about both.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiscardReason {
    /// Held for less than [`MINIMUM_HOLD`].
    TooShort,
    /// The cancel key.
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Command {
    Nothing,
    Start,
    Decode,
    Discard(DiscardReason),
}

impl State {
    pub fn indicator_phase(self) -> IndicatorPhase {
        match self {
            Self::Idle => IndicatorPhase::Idle,
            Self::Recording { .. } => IndicatorPhase::Recording,
            Self::Transcribing { .. } => IndicatorPhase::Transcribing,
        }
    }
    pub fn recording(self) -> bool {
        matches!(self, Self::Recording { .. })
    }
    pub fn latched(self) -> bool {
        matches!(self, Self::Recording { latched: true, .. })
    }
}

pub fn step(state: State, event: Event) -> (State, Command) {
    use Command::*;
    use Event::*;
    use State::*;
    match (state, event) {
        (Idle | Transcribing { .. }, Down { at, latch }) => (
            Recording {
                started: at,
                latched: latch,
            },
            Start,
        ),
        (
            Recording {
                started,
                latched: true,
            },
            Down { at, .. },
        )
        | (
            Recording {
                started,
                latched: false,
            },
            Up { at },
        ) => {
            if at.saturating_duration_since(started) < MINIMUM_HOLD {
                (Idle, Discard(DiscardReason::TooShort))
            } else {
                (
                    Transcribing {
                        started,
                        released: at,
                    },
                    Decode,
                )
            }
        }
        // Not a tap, however short: the user was speaking and the key they
        // would have released is gone. A latched recording has no other way
        // out at all — its key is already up, and the keyboard that carried
        // the cancel key left with it — so what was said is decoded.
        (Recording { started, .. }, HotkeyLost { at }) => (
            Transcribing {
                started,
                released: at,
            },
            Decode,
        ),
        (Recording { .. }, Cancel) => (Idle, Discard(DiscardReason::Cancelled)),
        // The audio is already captured and the release decode is about to
        // land. Escape is read from every keyboard regardless of focus, so
        // honouring it here would destroy a finished dictation.
        (Transcribing { .. }, Cancel) => (state, Nothing),
        (Transcribing { .. }, Finished) => (Idle, Nothing),
        _ => (state, Nothing),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use DiscardReason::*;

    #[test]
    fn all_state_event_pairs() {
        let t = Instant::now();
        let later = t + Duration::from_secs(1);
        let recording = |latched| State::Recording {
            started: t,
            latched,
        };
        let transcribing = State::Transcribing {
            started: t,
            released: later,
        };
        let started = State::Recording {
            started: later,
            latched: false,
        };
        let states = [State::Idle, recording(false), recording(true), transcribing];
        let events = [
            Event::Down {
                at: later,
                latch: false,
            },
            Event::Up { at: later },
            Event::HotkeyLost { at: later },
            Event::Finished,
            Event::Cancel,
        ];
        let decoded = State::Transcribing {
            started: t,
            released: later,
        };
        let expected = [
            [
                (started, Command::Start),
                (State::Idle, Command::Nothing),
                (State::Idle, Command::Nothing),
                (State::Idle, Command::Nothing),
                (State::Idle, Command::Nothing),
            ],
            [
                (recording(false), Command::Nothing),
                (decoded, Command::Decode),
                (decoded, Command::Decode),
                (recording(false), Command::Nothing),
                (State::Idle, Command::Discard(Cancelled)),
            ],
            [
                (decoded, Command::Decode),
                (recording(true), Command::Nothing),
                // The only way out of a latched recording whose keyboard left.
                (decoded, Command::Decode),
                (recording(true), Command::Nothing),
                (State::Idle, Command::Discard(Cancelled)),
            ],
            [
                (started, Command::Start),
                (transcribing, Command::Nothing),
                (transcribing, Command::Nothing),
                (State::Idle, Command::Nothing),
                // A cancel cannot destroy a dictation that is already captured.
                (transcribing, Command::Nothing),
            ],
        ];
        for (i, state) in states.into_iter().enumerate() {
            for (j, event) in events.into_iter().enumerate() {
                assert_eq!(step(state, event), expected[i][j], "state {i}, event {j}");
            }
        }
    }

    /// A device loss is not a tap: it ends the recording by decoding it, at
    /// any hold length and whether or not the key was latched.
    #[test]
    fn a_lost_hotkey_decodes_whatever_was_said() {
        let t = Instant::now();
        for latched in [false, true] {
            for ms in [0, 119, 500] {
                let at = t + Duration::from_millis(ms);
                assert_eq!(
                    step(
                        State::Recording {
                            started: t,
                            latched
                        },
                        Event::HotkeyLost { at }
                    ),
                    (
                        State::Transcribing {
                            started: t,
                            released: at
                        },
                        Command::Decode
                    ),
                    "latched {latched}, held {ms}ms"
                );
            }
        }
    }

    #[test]
    fn taps_and_latch_boundary() {
        let t = Instant::now();
        for latched in [false, true] {
            for ms in [0, 119, 120, 121] {
                let s = State::Recording {
                    started: t,
                    latched,
                };
                let at = t + Duration::from_millis(ms);
                let e = if latched {
                    Event::Down { at, latch: false }
                } else {
                    Event::Up { at }
                };
                let (state, command) = step(s, e);
                if ms < 120 {
                    assert_eq!(command, Command::Discard(TooShort));
                    assert_eq!(state, State::Idle);
                } else {
                    assert_eq!(command, Command::Decode);
                    assert_eq!(
                        state,
                        State::Transcribing {
                            started: t,
                            released: at
                        }
                    );
                }
            }
        }
    }
}
